Framework/Geometry/inc/MantidGeometry/Objects/BeamlineRayTracer.h

```diff
@@ -0,0 +1,335 @@
+        continue;
+      }
+
+      // Process a rectangular bank or normal component
+      if (childComponentType == Types::Rectangular) {
+        IntersectionHelpers::checkIntersectionWithRectangularBank(
+            track, componentInfo, i);
+      } else {
+        IntersectionHelpers::checkIntersectionWithComponent(track,
+                                                            componentInfo, i);
+      }
+    }
+  }
+}
+
+/**
+ * Return the results of any trace() calls since the last call to getResults.
+ *
+ * @return A collection of links defining intersection information
+ */
+Links getResults(Track &resultsTrack) {
+  // Create a list of the results
+  Links results(resultsTrack.cbegin(), resultsTrack.cend());
+
+  // Clear intersection results
+  resultsTrack.clearIntersectionResults();
+
+  return results;
+}
+
+/**
+ * Trace a given track from the source in the given direction.
+ *
+ * @param dir :: A directional vector. The starting point is defined by the
+ * instrument source.
+ * @param componentInfo :: The object used to access the source position.
+ * @return Links :: A collection of links defining intersection information.
+ */
+Links traceFromSource(const Kernel::V3D &dir,
+                      const ComponentInfo &componentInfo) {
+
+  // Create a results track
+  Track resultsTrack(componentInfo.sourcePosition(), dir);
+
+  // Fire the test ray at the instrument
+  fireRay(resultsTrack, componentInfo);
+
+  return getResults(resultsTrack);
+}
+
+/**
+ * Trace a given track from the sample position in the given direction.
+ *
+ * @param dir :: A directional vector. The starting point is defined by the
+ * source.
+ * @param componentInfo :: The object used to access the source position.
+ * @return Links :: A collection of links defining intersection information.
+ */
+Links traceFromSample(const Kernel::V3D &dir,
+                      const ComponentInfo &componentInfo) {
+
+  // Create a new results track
+  Track resultsTrack(componentInfo.samplePosition(), dir);
+
+  // Fire the test ray at the instrument
+  fireRay(resultsTrack, componentInfo);
+
+  return getResults(resultsTrack);
+}
+
+/**
+ * Gets the results of the trace, then returns the first detector
+ * index found in the results.
+ *
+ * @return size_t index value or throws an error if index is invalid
+ */
+size_t getDetectorResult(const ComponentInfo &componentInfo,
+                         Track &resultsTrack) {
+  // Store the results
+  Links results = getResults(resultsTrack);
+
+  // Go through all results via an iterator
+  Links::const_iterator resultIterator = results.begin();
+
+  // Return the first detectorIndex
+  while (resultIterator != results.end()) {
+    // Increment the iterator
+    ++resultIterator;
+
+    // Get the detector index
+    auto index = componentInfo.indexOf(resultIterator->componentID);
+
+    // Validate the index
+    if (index >= 0 && index <= componentInfo.size()) {
```

Comment:
Is this necessary. I think `componentInfo.indexOf` would throw if you tried to do a map lookup with an invalid key?